File: mio-archy-tool/src/main/java/com/miovision/Renderer/RenderInfoOrganization.java
package com.miovision.Renderer;

import com.miovision.Model.Job;
import com.miovision.Model.Mioarchy;
import com.miovision.Model.Organization;
import com.mxgraph.view.mxGraph;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by vleipnik on 2015-09-01.
 */
public class RenderInfoOrganization extends RenderInfo {
    private Organization org;
    private Mioarchy mioarchy;
    private boolean isLeaf;
    private RenderInfoCircles currentOrgContributorCircles;
    private RenderInfoCircles subOrgCircles;
    private int numSubOrgCircles;
    private List<RenderInfoOrganization> children;
    private List<Job> jobsAtThisLevel;
    private double maxOrgCircleDiameter;

    // constants
    final double CIRCLE_DIAMETER = 40;
    final double MIN_DISTANCE_BETWEEN_CIRCLES = 40;

    public RenderInfoOrganization(Organization org, Mioarchy mioarchy) {
        this.org = org;
        this.children = new ArrayList<>();
        this.mioarchy = mioarchy;

        // retrieve child orgs
        List<Organization> childOrgs = mioarchy.getOrganizationChildren(org);

        // a leaf org is rendered differently than a parent org
        this.isLeaf = childOrgs.size() == 0;

        // get list of jobs
        this.jobsAtThisLevel = mioarchy.getOrganizationJobs(org, false);

        // does org have children?
        if (this.isLeaf) {
            // no, treat this as a group of contributors - assume circle rendering
            // generate circle info for this org (it is a LEAF!)
            this.currentOrgContributorCircles = new RenderInfoCircles(
                    jobsAtThisLevel.size(), MIN_DISTANCE_BETWEEN_CIRCLES, CIRCLE_DIAMETER, true);

            // circle
            this.width = calculateBoundingCircleDiameter(
                    this.currentOrgContributorCircles.width(),
                    this.currentOrgContributorCircles.height());
            this.height = width;

            maxOrgCircleDiameter = this.width; // not really used, just being diligent
        } else {
            // yes, and populate child org infos
            childOrgs.forEach(o -> this.children.add(new RenderInfoOrganization(o, mioarchy)));

            // next, take the remaining nodes and treat them as a circle (without having explicitly an org)
            this.currentOrgContributorCircles = new RenderInfoCircles(
                    jobsAtThisLevel.size(), MIN_DISTANCE_BETWEEN_CIRCLES, CIRCLE_DIAMETER, true);

            //next, treat each sub org as a circle, and determine position based on the LARGEST circle (equidistant)
            //draw a circle around each sub org with radius = width / 2

            // determine the radius of the new circle based on width and height of the biggest sub org
            this.children.forEach(ri -> {
                maxOrgCircleDiameter = Math.max(maxOrgCircleDiameter, ri.width());
                maxOrgCircleDiameter = Math.max(maxOrgCircleDiameter, ri.height());
            });
            // also include the fake circle
            maxOrgCircleDiameter = Math.max(maxOrgCircleDiameter, this.currentOrgContributorCircles.width());
            maxOrgCircleDiameter = Math.max(maxOrgCircleDiameter, this.currentOrgContributorCircles.height());

            // include the org without a circle as a fake org
            numSubOrgCircles = childOrgs.size();
            //numSubOrgCircles = jobsAtThisLevel.size() == 0 ? childOrgs.size() : childOrgs.size() + 1;

            // calculate circle rendering info as though the sub orgs were just circles
            this.subOrgCircles = new RenderInfoCircles(
                    numSubOrgCircles, MIN_DISTANCE_BETWEEN_CIRCLES, maxOrgCircleDiameter, false);

            // max dimensions determine by the sub org circle max dimensions
            this.width = this.subOrgCircles.width();
            this.height = this.subOrgCircles.height();
        }
    }
    /*
    if node has no orgs inside it, render it as a circle
    second, render each sub-org, and get a size back
    remaining nodes in the org should be rendered as a mini circle, with size back
    now we have the sub orgs and fake sub org calculated
    next, treat each sub org as a circle, and determine position based on the LARGEST circle (equidistant)
    draw a circle around each sub org with radius = width / 2
    actually rendering based on render info applies rendering rules
     */
    public void render(double x, double y, mxGraph graph) {
        Point[] circleLocations;

        if (isLeaf) {
            // translate to 0,0
            Point d = getXYOffsetFromPoints(currentOrgContributorCircles.circleCenters());
            // account for differences in circle sizes
            double w = (this.width - currentOrgContributorCircles.width()) / 2;
            double h = (this.height - currentOrgContributorCircles.height()) / 2;
            // now account for containing circle center
            double dx = x - d.x + w;
            double dy = y - d.y + h;

            // now move all the circle locations as needed
            circleLocations = translatePoints(currentOrgContributorCircles.circleCenters(), dx, dy);

            // render jobs (circles)
            for (int i = 0; i < jobsAtThisLevel.size(); i++) {
                String defaultStyle = "shape=ellipse;whiteSpace=wrap;gradientColor=none";
                String color = determineContributorColor(jobsAtThisLevel.get(i), mioarchy);
                double defaultWidth = CIRCLE_DIAMETER;
                double defaultHeight = CIRCLE_DIAMETER;
                double cx = circleLocations[i].x;
                double cy = circleLocations[i].y;

                String label;
                if (jobsAtThisLevel.get(i).employee() != null) {
                    label = jobsAtThisLevel.get(i).employee().shortName().toLowerCase();
                } else {
                    label = "NEW"; // not yet hired
                }

                // actually "draw" :)
                graph.getModel().beginUpdate();
                try {
                    Object parent = graph.getDefaultParent();
                    graph.insertVertex(parent, null, label, cx, cy, defaultWidth, defaultHeight, defaultStyle +
                            ";gradientColor=" + color);
                } finally {
                    graph.getModel().endUpdate();
                }

                // now we will draw a circle around our org
                graph.getModel().beginUpdate();
                try {
                    Object parent = graph.getDefaultParent();
                    String orgLabel = org.name();
                    graph.insertVertex(parent, null, orgLabel, x, y, this.width, this.height,
                            "shape=ellipse;fillColor=none;whiteSpace=wrap;fillColor=none;" +
                                    "abelPosition=center;verticalLabelPosition=middle;align=center;verticalAlign=top;");
                } finally {
                    graph.getModel().endUpdate();
                }
            }

        } else {
            // note that this is an organization that has other organizations but also child jobs at this level
            boolean isParentOrgWithJobs = jobsAtThisLevel.size() > 0;

            // translate to 0,0
            Point d = getXYOffsetFromPoints(currentOrgContributorCircles.circleCenters());
            // now account for containing circle center, using the org sub circles as reference
            //TODO: this might be causing the offset problem for DES which has a smaller circle
            double dx = x - d.x + (this.width - this.subOrgCircles.width()) / 2;
            double dy = y - d.y + (this.height - this.subOrgCircles.height()) / 2;
            circleLocations = translatePoints(subOrgCircles.circleCenters(), dx, dy);

            // render organizations (circles)
            for (int i = 0; i < this.children.size(); i++) {
                // current org rendering info
                RenderInfoOrganization orgRenderInfo = this.children.get(i);
                String orgLabel = orgRenderInfo.org.name();

                graph.getModel().beginUpdate();
                try {
                    Object parent = graph.getDefaultParent();
                    graph.insertVertex(parent, null, orgLabel, circleLocations[i].x, circleLocations[i].y,
                            orgRenderInfo.width, orgRenderInfo.height,
                            "shape=ellipse;fillColor=none;whiteSpace=wrap;fillColor=none;" +
                            "abelPosition=center;verticalLabelPosition=middle;align=center;verticalAlign=top;");
                } finally {
                    graph.getModel().endUpdate();
                }

                // render organization internals
                orgRenderInfo.render(circleLocations[i].x, circleLocations[i].y, graph);
            }

            /*
            // now render the jobs at this level

            // translate to 0,0
            Point d = getXYOffsetFromPoints(currentOrgContributorCircles.circleCenters);

            double w = (this.width - currentOrgContributorCircles.width()) / 2;
            double h = (this.height - currentOrgContributorCircles.height()) / 2;
            if (isParentOrgWithJobs) {
                w = 0;
                h = 0;
            }

            // now account for containing circle center
            double dx = x - d.x + w;
            double dy = y - d.y + h;

            // if this the fake circle, we have to offset to account for the location the fake circle
            // this will be the very last circle in the list

            Point fakeCircleCenter = this.subOrgCircles.circleCenters[this.subOrgCircles.circleCenters.length - 1];
            dx += fakeCircleCenter.getX();
            dy += fakeCircleCenter.getY();

            // now move all the circle locations as needed
            Point[] circleLocations = translatePoints(currentOrgContributorCircles.circleCenters, dx, dy);

            // render jobs (circles)
            for (int i = 0; i < jobsAtThisLevel.size(); i++) {
                String defaultStyle = "shape=ellipse;whiteSpace=wrap;gradientColor=none";
                String color = determineContributorColor(jobsAtThisLevel.get(i), mioarchy);
                double defaultWidth = CIRCLE_DIAMETER;
                double defaultHeight = CIRCLE_DIAMETER;
                double cx = circleLocations[i].x;
                double cy = circleLocations[i].y;

                String label;
                if (jobsAtThisLevel.get(i).employee() != null) {
                    label = jobsAtThisLevel.get(i).employee().shortName().toLowerCase();
                } else {
                    label = "NEW"; // not yet hired
                }

                // actually "draw" :)
                graph.getModel().beginUpdate();
                try {
                    Object parent = graph.getDefaultParent();
                    graph.insertVertex(parent, null, label, cx, cy, defaultWidth, defaultHeight, defaultStyle +
                            ";gradientColor=" + color);
                } finally {
                    graph.getModel().endUpdate();
                }
            }
            */
        }
    }

    private String determineContributorColor(Job c, Mioarchy mioarchy)
    {
        String colorString = "";

        // this is a lead
        if (c.role().name().toLowerCase().indexOf("lead") >= 0) {
            colorString += "dark";
        }

        // by app now
        if (c.application() != null) {
            if (c.application().name().toLowerCase().indexOf("product") >= 0) {
                colorString = "green";
            } else if (c.application().name().toLowerCase().indexOf("strategic") >= 0) {
                colorString = "red";
            } else if (c.application().name().toLowerCase().indexOf("innovation") >= 0) {
                colorString = "gray";
            } else if (c.application().name().toLowerCase().indexOf("organizational development") >= 0) {
                colorString = "yellow";
            } else if (c.application().name().toLowerCase().indexOf("culture committee") >= 0) {
                colorString = "pink";
            } else if (c.application().name().toLowerCase().indexOf("quality control") >= 0) {
                colorString = "magenta";
            }
        } else {
            // try by organization
            if (c.organization() != null) {
                if (mioarchy.isDescendantOfOrganization(c.organization(), mioarchy.organizationByName("Engineering"))) {
                    colorString = "purple";
                } else if (mioarchy.isDescendantOfOrganization(c.organization(), mioarchy.organizationByName("Finance"))) {
                    colorString = "orange";
                } else if (mioarchy.isDescendantOfOrganization(c.organization(), mioarchy.organizationByName("Operations"))) {
                    colorString = "blue";
                }
            } else {
                colorString = "white";
            }
        }
        return colorString;
    }

    // Moves a list of points by dx,dy
    private Point[] translatePoints(Point[] points, double dx, double dy) {
        Point[] pointsTranslated = new Point[points.length];
        for (int i = 0; i < points.length; i++) {
            pointsTranslated[i] = new Point();
            pointsTranslated[i].setLocation(points[i].getX() + dx, points[i].getY() + dy);
        }
        return pointsTranslated;
    }

    private double calculateBoundingCircleDiameter(double width, double height) {
        return Math.sqrt(width*width + height*height);
    }

    // TODO: this ignores circle radius -- this could be causing positioning errors i've been seeing
    private Point getXYOffsetFromPoints(Point[] points) {
        double x = Double.MAX_VALUE;
        double y = Double.MAX_VALUE;
        for (int i = 0; i < points.length; i++) {
            x = Math.min(x, points[i].x);
            y = Math.min(y, points[i].y);
        }
        Point p = new Point();
        p.setLocation(x, y);
        return p;
    }
}

File: mio-archy-tool-js/lib/mioarchy/render/RenderInfoOrganization.js
'use strict';

function RenderInfoOrganization(org, mioarchy)
{
    // constants
    this.CIRCLE_DIAMETER = 40;
    this.MIN_DISTANCE_BETWEEN_CIRCLES = 40;

    this.org = org;
    this.childRenderingInfos = [];
    this.mioarchy = mioarchy;
    this.orgLevel = mioarchy.getOrganizationLevel( org );
    this.numJobsInTotal = mioarchy.getOrganizationJobs( org, true ).length;

    // retrieve child orgs
    this.childOrgs = mioarchy.getOrganizationChildren( org );

    // a leaf org is rendered differently than a parent org
    this.isLeaf = this.childOrgs.length == 0;

    // get list of jobs
    this.jobsAtThisLevel = mioarchy.getOrganizationJobs( org, false );
    this.jobsAtThisLevelPosition = { x: 0, y: 0 };

    // yes, and populate child org infos
    for (var o in this.childOrgs) {
        this.childRenderingInfos.push( new RenderInfoOrganization( this.childOrgs[o], mioarchy ) );
    }

    // the remaining contributors at this level could be treated as their own organization, without the
    // actual org boundaries, so essentially they will be accounted for as a circle in the rendering calculations
    // but will not have a circle drawn around them
    var useMiddle = this.childOrgs.length > 0;
    this.circleForContributorsAtThisOrgLevel = new RenderInfoCircles(this.jobsAtThisLevel.length,
        this.MIN_DISTANCE_BETWEEN_CIRCLES, this.CIRCLE_DIAMETER, useMiddle); // orgs with children use middle

    determineMaximumSubOrgDimensions.call(this);

    // include the org without a circle as a fake org
    this.numSubOrgCircles = this.jobsAtThisLevel.length == 0 ? this.childOrgs.length : this.childOrgs.length + 1;

    if (this.orgLevel > 2 && !this.org.isApplication) {
        processCircularSubOrgRendering.call(this);
    } else {
        processRectangularOrgRendering.call(this);
    }
}

function determineMaximumSubOrgDimensions() {
    // determine the radius of the new circle based on width and height of the biggest sub org
    this.maxOrgWidth = 0;
    this.maxOrgHeight = 0;
    for (var riIndex in this.childRenderingInfos) {
        var ri = this.childRenderingInfos[riIndex];
        this.maxOrgWidth = Math.max(this.maxOrgWidth, ri.width);
        this.maxOrgHeight = Math.max(this.maxOrgHeight, ri.height);
    }

    // also include the fake circle
    this.maxOrgCircleDiameter = Math.max( this.maxOrgWidth, this.maxOrgHeight );
    this.maxOrgCircleDiameter = Math.max( this.maxOrgCircleDiameter, this.circleForContributorsAtThisOrgLevel.width );
    this.maxOrgCircleDiameter = Math.max( this.maxOrgCircleDiameter, this.circleForContributorsAtThisOrgLevel.height );
}

function processCircularSubOrgRendering() {

    // calculate circle rendering info as though the sub orgs were just circles
    var useMiddle = this.childOrgs.length > 0;
    this.subOrgCircles = new RenderInfoCircles(this.numSubOrgCircles, this.MIN_DISTANCE_BETWEEN_CIRCLES, this.maxOrgCircleDiameter, useMiddle);

    // max dimensions determine by the sub org circle max dimensions
    this.width = this.subOrgCircles.width + this.MIN_DISTANCE_BETWEEN_CIRCLES;
    this.height = this.subOrgCircles.height + this.MIN_DISTANCE_BETWEEN_CIRCLES;
}

function processRectangularOrgRendering()
{
    // we will be sorting orgs into two piles: normal orgs and applications
    var applicationOrgInfos = { positions:[], orgRefs:[], width:0, height:0, cx:this.MIN_DISTANCE_BETWEEN_CIRCLES };
    var normalOrgInfos = { positions:[], orgRefs:[], width:0, height:0, cx:this.MIN_DISTANCE_BETWEEN_CIRCLES };

    // first, find the max heights so we can properly calculate the y positions and total org height
    for (var r in this.childRenderingInfos) {
        var ri = this.childRenderingInfos[r];
        if (ri.org.isApplication) {
            applicationOrgInfos.height = Math.max(applicationOrgInfos.height, ri.height);
        } else {
            normalOrgInfos.height = Math.max(normalOrgInfos.height, ri.height);
        }
    }
    // also include the contributor org in the set of apps
    if (this.jobsAtThisLevel.length > 0) {
        applicationOrgInfos.height = Math.max(applicationOrgInfos.height, this.circleForContributorsAtThisOrgLevel.height);
    }

    // we are now ready to determine the height of this org
    // (m + hmax_a + 2m + hmax_o + m)
    this.height = applicationOrgInfos.height + normalOrgInfos.height + this.MIN_DISTANCE_BETWEEN_CIRCLES * 4;

    // add the contributor circles first in the set of apps if it exists
    if (this.jobsAtThisLevel.length > 0) {
        // org must be corrected to 0,0 (TODO: this really should be circle org rendering's responsibility!)
        var dx = this.circleForContributorsAtThisOrgLevel.width/2;
        var dy = this.circleForContributorsAtThisOrgLevel.height/2;

        applicationOrgInfos.cy = this.MIN_DISTANCE_BETWEEN_CIRCLES + (applicationOrgInfos.height-this.circleForContributorsAtThisOrgLevel.height)/2
        this.jobsAtThisLevelPosition = {x: applicationOrgInfos.cx + dx, y: applicationOrgInfos.cy + dy };
        applicationOrgInfos.cx += this.circleForContributorsAtThisOrgLevel.width + this.MIN_DISTANCE_BETWEEN_CIRCLES;
        applicationOrgInfos.width = applicationOrgInfos.cx;
    }

    // next, calculate the x and y positions for the apps and normal sub orgs
    for (var r in this.childRenderingInfos) {
        var ri = this.childRenderingInfos[r];
        if (ri.org.isApplication) {
            applicationOrgInfos.cy = this.MIN_DISTANCE_BETWEEN_CIRCLES + (applicationOrgInfos.height-ri.height)/2;
            applicationOrgInfos.positions.push( {x: applicationOrgInfos.cx, y: applicationOrgInfos.cy, } );
            applicationOrgInfos.orgRefs.push( ri.org );
            applicationOrgInfos.cx += ri.width + this.MIN_DISTANCE_BETWEEN_CIRCLES;
            applicationOrgInfos.width = applicationOrgInfos.cx;
        } else {
            normalOrgInfos.cy = this.MIN_DISTANCE_BETWEEN_CIRCLES + (normalOrgInfos.height-ri.height)/2;
            normalOrgInfos.positions.push( {x: normalOrgInfos.cx, y: normalOrgInfos.cy } );
            normalOrgInfos.orgRefs.push( ri.org );
            normalOrgInfos.cx += ri.width + this.MIN_DISTANCE_BETWEEN_CIRCLES;
            normalOrgInfos.width = normalOrgInfos.cx;
        }
    }
    // finally, determine which is the widest, and that becomes the total org width
    this.width = Math.max( applicationOrgInfos.cx, normalOrgInfos.cx );
    // the last step is to translate each coord to its respective position
    if (applicationOrgInfos.width < normalOrgInfos.width) {
        // applications are smaller, so translate the application orgs on the x axis (centering)
        var dx = (this.width - applicationOrgInfos.width) / 2;
        applicationOrgInfos.positions = RenderingUtilities.translatePoints(applicationOrgInfos.positions, dx, 0);
    } else {
        // normal orgs are wider, so translate them on the x axis (centering)
        var dx = (this.width - normalOrgInfos.width) / 2;
        normalOrgInfos.positions = RenderingUtilities.translatePoints(normalOrgInfos.positions, dx, 0);
    }
    // move the contributor circle over
    this.jobsAtThisLevelPosition.x += (this.width - applicationOrgInfos.width) / 2;
    // also, normal orgs always get rendered underneath the apps
    var dy = applicationOrgInfos.height + this.MIN_DISTANCE_BETWEEN_CIRCLES * 2;
    normalOrgInfos.positions = RenderingUtilities.translatePoints(normalOrgInfos.positions, 0, dy);

    // now combine both arrays into the final product
    this.subOrgPositions = [];
    for (var i = 0; i < applicationOrgInfos.positions.length; i++) {
        var position = applicationOrgInfos.positions[i];
        var orgName = applicationOrgInfos.orgRefs[i].name;
        this.subOrgPositions[orgName] = position;
    }
    for (var i = 0; i < normalOrgInfos.positions.length; i++) {
        var position = normalOrgInfos.positions[i];
        var orgName = normalOrgInfos.orgRefs[i].name;
        this.subOrgPositions[orgName] = position;
    }

    // final correction: if there weren't any normal orgs, i.e. this was an app, reduce the overall height
    if (normalOrgInfos.positions.length == 0) {
        this.height -= this.MIN_DISTANCE_BETWEEN_CIRCLES * 2;
    }
}

function renderJobsAtThisOrgLevel(x, y, dx, dy, graph) {
    // now move all the circle locations as needed
    var circleLocations = RenderingUtilities.translatePoints(this.circleForContributorsAtThisOrgLevel.circleCenters, dx, dy);

    // render jobs (circles)
    for (var j in this.jobsAtThisLevel) {
        var job = this.mioarchy.jobs[this.jobsAtThisLevel[j]];

        var defaultStyle = "shape=ellipse;whiteSpace=wrap;gradientColor=none;editable=0;";
        var color = this.determineContributorColor(job, this.mioarchy);
        var defaultWidth = this.CIRCLE_DIAMETER;
        var defaultHeight = this.CIRCLE_DIAMETER;

        // normal coords are the x,y coords + the circle coords
        // (the normal drawing is the top-left of the circle, so we must offset by the circle dimensions)
        var cx = x + circleLocations[j].x - this.CIRCLE_DIAMETER / 2;
        var cy = y + circleLocations[j].y - this.CIRCLE_DIAMETER / 2;

        var label;
        if (job.contributor && this.mioarchy.contributors[job.contributor]) {
            label = this.mioarchy.namesToShortNames[job.contributor].shortName;
        } else {
            label = "NEW"; // not yet hired
        }

        // actually "draw" :)
        graph.getModel().beginUpdate();
        try {
            var parent = graph.getDefaultParent();
            var v = graph.insertVertex(parent, null, label, cx, cy, defaultWidth, defaultHeight,
                defaultStyle + "gradientColor=" + color);
            // attach the org info to the vertex
            v.mioObject = job;
            this.mioarchy.jobToVertex[job.id] = v;
        } finally {
            graph.getModel().endUpdate();
        }
    }
}

RenderInfoOrganization.prototype =
{
    /*
     if node has no orgs inside it, render it as a circle
     second, render each sub-org, and get a size back
     remaining nodes in the org should be rendered as a mini circle, with size back
     now we have the sub orgs and fake sub org calculated
     next, treat each sub org as a circle, and determine position based on the LARGEST circle (equidistant)
     draw a circle around each sub org with radius = width / 2
     actually rendering based on render info applies rendering rules
     */
    render: function(x, y, graph) {
        /*
        if (this.jobsAtThisLevel == 0) {
            console.log(this.org.name + " has no contributors. Skipping rendering.");
            return;
        }*/
        //==============================================================================================================
        // TOP LEVEL ORGANIZATIONS
        if (this.orgLevel <= 2 || this.org.isApplication) {
            // if there are jobs, render them
            if (this.jobsAtThisLevel.length > 0) {
                if (this.subOrgPositions) {
                    var dx, dy;
                    cx = x + this.jobsAtThisLevelPosition.x;
                    cy = y + this.jobsAtThisLevelPosition.y;
                    renderJobsAtThisOrgLevel.call(this, cx, cy, 0, 0, graph);
                }
            }
            //==========================================================================================================
            // CHILDREN OF TOP LEVEL ORGANIZATION
            for (var i = 0; i < this.childRenderingInfos.length; i++) {
                // current org rendering info
                var orgRenderInfo = this.childRenderingInfos[i];
                var orgPosition = this.subOrgPositions[ orgRenderInfo.org.name ];
                if (typeof(orgPosition) == 'undefined') {
                    console.log("UNDEFINED ORG POSITION!!!");
                }
                // render organization internals
                orgRenderInfo.render(x + orgPosition.x, y + orgPosition.y, graph);
            }
            //==========================================================================================================
            // BOUNDING BOX FOR TOP LEVEL ORGANIZATION
            graph.getModel().beginUpdate();
            try {
                var parent = graph.getDefaultParent();
                var orgLabel = this.org.name;

                // stroke is different for applications
                var stroke = "";
                var shape = "rounded=1;"
                if (this.org.isApplication) {
                    // highlight the jobs this person has taken on
                    var appColor = this.mioarchy.applications[this.org.name].color;
                    stroke = "strokeWidth=10;strokeColor="+appColor+";opacity=100;";
                    shape = "shape=process;";
                }
                var vertex = graph.insertVertex(parent, null, orgLabel, x, y, this.width, this.height,
                    shape + "fillColor=none;whiteSpace=wrap;editable=0;" + stroke +
                    "labelPosition=center;verticalLabelPosition=middle;align=center;verticalAlign=top;");
                // attach the org info to the vertex
                vertex.mioObject = this.org;
                this.mioarchy.orgToVertex[orgLabel] = vertex;
                // send to back
                graph.cellsOrdered([vertex], true);

            } finally {
                graph.getModel().endUpdate();
            }
        }
        //==============================================================================================================
        // SUB ORGANIZATION
        else {
            //==========================================================================================================
            // SUB ORGANIZATION WITH CHILDREN
            if (!this.isLeaf) {
                // first we will draw the circle that will contain our sub circles
                graph.getModel().beginUpdate();
                try {
                    var parent = graph.getDefaultParent();
                    var orgLabel = this.org.name;

                    // the normal drawing is the top-left of the circle, so we must offset by the circle dimensions
                    var cx = x;
                    var cy = y;

                    var vertex = graph.insertVertex(parent, null, orgLabel, cx, cy, this.width, this.height,
                        "shape=ellipse;fillColor=none;whiteSpace=wrap;editable=0;" +
                        "labelPosition=center;verticalLabelPosition=middle;align=center;verticalAlign=top;");
                    // attach the org info to the vertex
                    vertex.mioObject = this.org;
                    this.mioarchy.orgToVertex[orgLabel] = vertex;
                    // send to back
                    graph.cellsOrdered([vertex], true);

                } finally {
                    graph.getModel().endUpdate();
                }
                // render organizations (circles)
                var circleLocations = this.subOrgCircles.circleCenters;
                for (var i = 0; i < this.childRenderingInfos.length; i++) {
                    // current org rendering info
                    var orgRenderInfo = this.childRenderingInfos[i];

                    // offset by containing circle dimensions, as well as sub circle dimensions
                    var cx = x + circleLocations[i].x + this.width / 2 - orgRenderInfo.width / 2;
                    var cy = y + circleLocations[i].y + this.width / 2 - orgRenderInfo.height / 2;

                    // render organization internals
                    orgRenderInfo.render(cx, cy, graph);
                }
            }
            //==========================================================================================================
            // LEAF ORGANIZATION
            var dx = 0;
            var dy = 0;
            // only draw the containing circle if this is a leaf
            if (this.isLeaf) {
                // first we will draw the circle that will contain our job circles
                graph.getModel().beginUpdate();
                try {
                    var parent = graph.getDefaultParent();
                    var orgLabel = this.org.name;

                    // the normal drawing is the top-left of the circle, so we must offset by the circle dimensions
                    var cx = x;
                    var cy = y;

                    var vertex = graph.insertVertex(parent, null, orgLabel, cx, cy, this.width, this.height,
                        "shape=ellipse;fillColor=none;whiteSpace=wrap;editable=0;" +
                        "abelPosition=center;verticalLabelPosition=middle;align=center;verticalAlign=top;");
                    // attach the org info to the vertex
                    vertex.mioObject = this.org;
                    this.mioarchy.orgToVertex[orgLabel] = vertex;

                } finally {
                    graph.getModel().endUpdate();
                }
                // offset by half of the containing circle dimensions
                dx = this.width / 2;
                dy = this.height / 2;
            } else {
                // this is the last "virtual" circle in the calculate rendering infos for this org, which contains the
                // jobs at this level
                var lastCircle = this.subOrgCircles.circleCenters[this.subOrgCircles.circleCenters.length - 1];
                dx = lastCircle.x + this.width / 2;
                dy = lastCircle.y + this.width / 2;
            }
            //==========================================================================================================
            // JOBS AT THIS SUB ORGANIZATION LEVEL
            if (this.jobsAtThisLevel.length > 0) {
                // now move all the circle locations as needed
                var circleLocations = RenderingUtilities.translatePoints(this.circleForContributorsAtThisOrgLevel.circleCenters, dx, dy);

                // render jobs (circles)
                for (var j in this.jobsAtThisLevel) {
                    var job = this.mioarchy.jobs[this.jobsAtThisLevel[j]];

                    var defaultStyle = "shape=ellipse;whiteSpace=wrap;gradientColor=none;editable=0;";
                    var color = this.determineContributorColor(job, this.mioarchy);
                    var defaultWidth = this.CIRCLE_DIAMETER;
                    var defaultHeight = this.CIRCLE_DIAMETER;

                    // normal coords are the x,y coords + the circle coords
                    // (the normal drawing is the top-left of the circle, so we must offset by the circle dimensions)
                    var cx = x + circleLocations[j].x - this.CIRCLE_DIAMETER / 2;
                    var cy = y + circleLocations[j].y - this.CIRCLE_DIAMETER / 2;

                    var label;
                    if (job.contributor && this.mioarchy.contributors[job.contributor]) {
                        label = this.mioarchy.namesToShortNames[job.contributor].shortName;
                    } else {
                        label = "NEW"; // not yet hired
                    }

                    // actually "draw" :)
                    graph.getModel().beginUpdate();
                    try {
                        var parent = graph.getDefaultParent();
                        var v = graph.insertVertex(parent, null, label, cx, cy, defaultWidth, defaultHeight,
                            defaultStyle + ";gradientColor=" + color);
                        // attach the org info to the vertex
                        v.mioObject = job;
                        this.mioarchy.jobToVertex[job.id] = v;
                    } finally {
                        graph.getModel().endUpdate();
                    }
                }
            }
        }
    },
    determineContributorColor: function(job, mioarchy)
    {
        var colorString = "";

        // by app now
        if (job.application) {
            colorString = mioarchy.applications[job.application].color;
        } else {
            // try by organization
            if (job.organization) {
                var org = mioarchy.organizations[job.organization];
                if (mioarchy.isDescendantOfOrganization( org, mioarchy.organizations["Engineering"] )) {
                    colorString += "purple";
                } else if (mioarchy.isDescendantOfOrganization( org, mioarchy.organizations["Finance"] )) {
                    colorString += "orange";
                } else if (mioarchy.isDescendantOfOrganization( org, mioarchy.organizations["Operations"] )) {
                    colorString += "blue";
                } else if (mioarchy.isDescendantOfOrganization( org, mioarchy.organizations["Operations"] )) {
                    colorString += "blue";
                }
            } else {
                colorString = "white";
            }
        }
        return colorString;
    },

}